pandas/core/index.py

```diff
@@ -3260,8 +3260,12 @@ def _evaluate_compare(self, other):
                 elif isinstance(other, Index):
                     other = self._create_categorical(self, other.values, categories=self.categories, ordered=self.ordered)
 
+                if isinstance(other, (ABCCategorical, np.ndarray, ABCSeries)):
+                    if len(self.values) != len(other):
+                        raise ValueError("Lengths must match to compare")
+
                 if isinstance(other, ABCCategorical):
-                    if not (self.values.is_dtype_equal(other) and len(self.values) == len(other)):
```

Comment:
@jreback the unit test at line 1980 below hits this case. 

Currently (without this PR) we have this:
```
In [2]: pd.CategoricalIndex(['a', 'b']) == pd.Index(['a','b','c'])
TypeError: categorical index comparisions must have the same categories and ordered attributes
```
With this PR we have:
```
In [2]: pd.CategoricalIndex(['a', 'b']) == pd.Index(['a','b','c'])
ValueError: Lengths must match to compare
```